api/src/components/crosslink/service.ts

```diff
@@ -189,187 +189,252 @@ export const getVote = (crosslinkId: string, userId: string) =>
         }
     });
 
-export const getPublicationCrosslinks = async (publicationId: string, options?: I.GetPublicationCrosslinksOptions) => {
-    const { order, search, limit, offset, userIdFilter } = options || {};
-    const publicationInclude = {
+type GetPublicationCrosslinksQueryResult = {
+    id: string;
+    linkedPublicationId: string;
+    createdBy: string;
+    createdAt: string;
+    score: number;
+    linkedPublicationLatestLiveVersionId: string;
+    linkedPublicationTitle: string;
+    linkedPublicationPublishedDate: string;
+    linkedPublicationAuthorId: string;
+    linkedPublicationAuthorFirstName: string;
+    linkedPublicationAuthorLastName: string;
+};
+
+type RelativeCrosslink = {
+    id: string;
+    score: number;
+    createdBy: string;
+    createdAt: string;
+    linkedPublication: Pick<I.Publication, 'id'> & {
+        latestLiveVersion: Pick<I.PublicationVersion, 'title'> & {
+            user: Pick<I.User, 'id' | 'firstName' | 'lastName'>;
+        };
+    };
+};
+interface GetPublicationCrosslinksQueryOptions extends I.GetPublicationCrosslinksOptions {
+    excludedPublicationIds?: string[];
+}
+
+const getPublicationCrosslinksQuery = async (
+    publicationId: string,
+    options?: GetPublicationCrosslinksQueryOptions
+): Promise<{
+    results: RelativeCrosslink[];
+    total: number;
+}> => {
+    const { order, search, limit, offset, userIdFilter, excludedPublicationIds } = options || {};
+
+    const conditionalFilterClauses: Prisma.Sql[] = [];
+    // Some conditional filter clauses have to be tailored depending whether we are getting
+    // crosslinks "from" or "to" the publication.
+    const conditionalFilterClausesFrom: Prisma.Sql[] = [];
+    const conditionalFilterClausesTo: Prisma.Sql[] = [];
+
+    if (userIdFilter) {
+        conditionalFilterClauses.push(Prisma.sql`c."createdBy" = ${userIdFilter}`);
+    }
+
+    if (search) {
+        conditionalFilterClauses.push(Prisma.sql`to_tsvector('english', pv.title) @@ to_tsquery('english', ${search})`);
+    }
+
+    if (excludedPublicationIds?.length) {
+        conditionalFilterClausesTo.push(
+            Prisma.sql`c."publicationToId" NOT IN (${Prisma.join(excludedPublicationIds)})`
+        );
+        conditionalFilterClausesFrom.push(
+            Prisma.sql`c."publicationFromId" NOT IN (${Prisma.join(excludedPublicationIds)})`
+        );
+    }
+
+    const conditionalWhereTo =
```

Comment:
I wonder if all this conditional stuff lends to stored procedures being a better solution? This is quite difficult to read and follow. I've had a similar style of query in a project I worked on in the past, it was a bit of a nightmare trying to build the query up to test it outside of the api.